### ISSUE TITLE
Local cache doesn't cleanup properly

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#864]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/864)
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Jan 15, 2019, 09:31

## Summary

[//]: # (Summarize the bug encountered concisely)

Trying to build a project with no maximum size for the cache (hence defaults), I get:

```
[bschubert[[Gitlab user @me]](https://gitlab.com/me) $MYELEMENT]$ cat eb0beee1-build.31021.log
[--:--:--] START   MYELEMENT.bst: Build
[--:--:--] LOG     MYELEMENT.bst: Build environment for element MYELEMENT.bst

    HOME: /tmp
    LC_ALL: C
    LOGNAME: tomjon
    PATH: /usr/bin:/bin:/usr/sbin:/sbin
    SHELL: /bin/sh
    SOURCE_DATE_EPOCH: '1320937200'
    TERM: dumb
    TZ: UTC
    USER: tomjon
    USERNAME: tomjon
[--:--:--] BUG     MYELEMENT.bst: Build

    An unhandled exception occured:

    Traceback (most recent call last):
      File "/home/bschubert/.local/lib/python3.6/site-packages/buildstream/_scheduler/jobs/job.py", line 417, in _child_action
        result = self.child_process()  # pylint: disable=assignment-from-no-return
      File "/home/bschubert/.local/lib/python3.6/site-packages/buildstream/_scheduler/jobs/elementjob.py", line 94, in child_process
        return self._action_cb(self._element)
      File "/home/bschubert/.local/lib/python3.6/site-packages/buildstream/_scheduler/queues/buildqueue.py", line 71, in process
        return element._assemble()
      File "/home/1/.local/lib/python3.6/site-packages/buildstream/element.py", line 1572, in _assemble
        self.__sandbox(rootdir, output_file, output_file, self.__sandbox_config) as sandbox:  # nopep8
      File "/opt/rh/rh-python36/root/usr/lib64/python3.6/contextlib.py", line 81, in __enter__
        return next(self.gen)
      File "/home/bschubert/.local/lib/python3.6/site-packages/buildstream/element.py", line 2252, in __sandbox
        allow_real_directory=not self.BST_VIRTUAL_DIRECTORY)
      File "/home/bschubert/.local/lib/python3.6/site-packages/buildstream/_platform/linux.py", line 69, in create_sandbox
        return self._create_bwrap_sandbox(*args, **kwargs)
      File "/home/bschubert/.local/lib/python3.6/site-packages/buildstream/_platform/linux.py", line 130, in _create_bwrap_sandbox
        return SandboxBwrap(*args, **kwargs)
      File "/home/bschubert/.local/lib/python3.6/site-packages/buildstream/sandbox/_sandboxbwrap.py", line 56, in __init__
        super().__init__(*args, **kwargs)
      File "/home/bschubert/.local/lib/python3.6/site-packages/buildstream/sandbox/sandbox.py", line 145, in __init__
        os.makedirs(directory_, exist_ok=True)
      File "/opt/rh/rh-python36/root/usr/lib64/python3.6/os.py", line 220, in makedirs
        mkdir(name, mode)
    OSError: [Errno 28] No space left on device: '/home/bschubert/.cache/buildstream/build/rmtfetickpublish-librmtfetickpublish-dev-f97w5sj6/root'
```

## Steps to reproduce

[//]: # (How one can reproduce the issue - this is very important)
- Create a small partition to use as your cache
- Build a project that will fill it up

## What is the current bug behavior?

[//]: # (What actually happens)
The cache cleaning doesn't kick in

## What is the expected correct behavior?

[//]: # (What you should see instead)
Old entries in the cache should be removed

## Relevant logs and/or screenshots

[//]: # (Paste any relevant logs - please use code blocks ``` to format console output, logs, and code as it's hard to read otherwise. You can also add a Snippet and link it here. Check the markdown giude on Gitlab for further tips)

## Possible fixes

[//]: # (If you can, link to the line of code that might be responsible for the problem)

## Other relevant information

[//]: # (Add the BuildStream version. Substitute below the "x" by "1", "2" or the right version. Add more than one version if necessary)

* BuildStream version affected: /milestone %BuildStream_v1.x

----

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jennis]](https://gitlab.com/jennis) on Jan 15, 2019, 10:07

Hi [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert), exactly how big is this 'small' partition you've created? I wonder whether this has anything to do with the expected 2GB buffer.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Jan 15, 2019, 10:14

It was around 1 TB ;) and was only filled up after a few runs

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jan 19, 2019, 16:33

This is related to #660

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jan 19, 2019, 16:33

marked this issue as related to #660

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Feb 27, 2019, 14:51

Since various fixes landed:
* ce01f87e3 Ensures that at startup cache size is checked and a cleanup happens before anything else
* 22b3a0c18 Also ensures that we consult remaining disk space as well as quota usage in order to trigger cleanup jobs even if the quota is not exceeded but not much space remains

I think that the only parts of this issue which remain are crashes which happen when running out of physical disk space, and that we just have to handle those more gracefully.

In other words, the remaining parts of this are a duplicate of #660.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Feb 27, 2019, 14:56

agreed. I still want to test the changes "for real" and will close this isse once done

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Feb 27, 2019, 14:56

assigned to [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Mar 22, 2019, 14:12

closed
